doc/SERVICES.md

```diff
@@ -0,0 +1,227 @@
+# Services
+
+Libp2p ships with very little functionality by default, this is to allow the greatest amount of flexibility and to ensure, for example, if you are deploying to web browsers you only pull in the code that your application needs.
+
+The functionality of your Libp2p node can be extended by configuring additional services.
+
+```ts
+import { createLibp2p } from 'libp2p'
+import { identify } from '@libp2p/identify'
+
+const node = await createLibp2p({
+  //.. other config here
+  services: {
+    identify: identify()
+  }
+})
+```
+
+You can extend the capabilities of your node to suit your needs by writing custom services.
+
+## Writing custom services
+
+At it's simplest a service might look like this:
+
+```ts
+import { createLibp2p } from 'libp2p'
+
+function myService () {
+  return () => {
+    return {
+      saySomething: () => {
+        return 'Hello'
+      }
+    }
+  }
+}
+
+const node = await createLibp2p({
+  //.. other config here
+  services: {
+    myService: myService()
+  }
+})
+
+console.info(node.services.myService.saySomething()) // 'Hello'
+```
+
+### Init args
+
+Your service can take arguments that allow for custom config:
+
+> [!TIP]
+> Make all arguments optional with sensible defaults
+
+```ts
+import { createLibp2p } from 'libp2p'
+
+interface MyServiceInit {
+  message?: string
+}
+
+function myService (init: MyServiceInit = {}) {
+  return () => {
+    return {
+      saySomething: () => {
+        return init.message ?? 'Hello'
+      }
+    }
+  }
+}
+
+const node = await createLibp2p({
+  //.. other config here
+  services: {
+    myService: myService({
+      message: 'World'
+    })
+  }
+})
+
+console.info(node.services.myService.saySomething()) // 'World'
+```
+
+### Accessing libp2p components
+
+Services can access internal libp2p components such as the address manger and connection manager by accepting an argument to the returned function.
+
+> [!TIP]
+> Use an interface to limit the scope of your service's component requirements
+
+```ts
+import { createLibp2p } from 'libp2p'
+import type { PeerId } from '@libp2p/interface'
+
+interface MyServiceInit {
+  message?: string
+}
+
+interface MyServiceComponents {
+  peerId: PeerId
+}
+
+function myService (init: MyServiceInit = {}) {
+  return (components: MyServiceComponents) => {
+    return {
+      saySomething: () => {
+        return `${init.message ?? 'Hello, my peer id is'} ${components.peerId}`
+      }
+    }
+  }
+}
+
+const node = await createLibp2p({
+  //.. other config here
+  services: {
+    myService: myService()
+  }
+})
+
+console.info(node.services.myService.saySomething()) // 'Hello, my peer id is 123Koo....'
+```
+
+### Using classes
+
+Use a class for your service and returning an instance will allow you to keep your code manageable.
```

Comment:
I'm not sure it's appropriate to suggest a particular programming paradigms here, it seems beyond the scope of documenting configuration.

Reply:
Most people will just copy/paste so I think it's fair to show them the "most idiomatic" way of doing things, though you are right it's a bit weird to call this out specifically.

I've updated the examples.